meson.build

```diff
@@ -0,0 +1,301 @@
+    'libdir=/usr/lib64']
+  )
+# can't set sbindir=/sbin (vs /usr/sbin) here, since that only
+# works on newer version of meson (>0.63.0)
+
+# needed for path checking
+fs = import('fs')
+
+# handle options
+no_systemd = get_option('no_systemd')
+home_dir = get_option('homedir')
+if not fs.is_absolute(home_dir)
+  home_dir = get_option('sysconfdir') / home_dir
+endif
+db_root = get_option('dbroot')
+if not fs.is_absolute(db_root)
+  db_root = get_option('sysconfdir') / db_root
+endif
+rules_dir = get_option('rulesdir')
+if not fs.is_absolute(rules_dir)
+  rules_dir = get_option('sysconfdir') / rules_dir
+endif
+if not no_systemd
+  systemd_dir = get_option('systemddir')
+  unit_dir = systemd_dir / 'system'
+  generator_dir = systemd_dir /'system-generators'
+endif
+c_args = get_option('c_args')
+man_dir = get_option('mandir')
+iscsi_sbindir = get_option('iscsi_sbindir')
+
+#
+# get list of sources from subdirs (iscsiuio included at bottom)
+#
+subdir('sysdeps')
+subdir('usr')
+subdir('libopeniscsiusr')
+subdir('utils')
+subdir('etc')
+subdir('doc')
+
+conf = configuration_data()
+conf.set('SBINDIR', iscsi_sbindir)
+conf.set('HOMEDIR', home_dir)
+conf.set('DBROOT', db_root)
+
+
+# set up general C args
+genl_cargs = [
+  '-D_GNU_SOURCE',
+  '-DOFFLOAD_BOOT_SUPPORTED',
+  '-DISCSI_DB_ROOT="@0@"'.format(db_root),
+  '-DLOCK_DIR="@0@"'.format(home_dir),
+  '-DISCSI_VERSION_STR="@0@"'.format(meson.project_version())]
+
+# set up no-system flag, if needed
+if no_systemd
+  genl_cargs += '-DNO_SYSTEMD'
+endif
+
+#
+# set up include directories
+#
+main_inc = include_directories('include')
+lib_inc = include_directories('libopeniscsiusr')
+usr_inc = include_directories('usr')
+
+
+# set up dependencies
+kmod_dep = dependency('libkmod')
+crypto_dep = dependency('libcrypto')
+mount_dep = dependency('mount')
+if not no_systemd
+  systemd_dep = dependency('libsystemd')
+endif
+
+#
+# build two static libs that programs in 'usr' will need, which
+# do not depend on anything local
+#
+libsysdeps = static_library('sysdeps', sysdeps_files,
+  include_directories: [usr_inc, main_inc, lib_inc],
+  c_args: '-Wno-all')
+libfwparam = static_library('fwparam', fwparam_files,
+  include_directories: [usr_inc, main_inc, lib_inc],
+  c_args: [genl_cargs, '-Wno-all'])
+
+#
+# build libopeniscsiusr, since it does not
+# depend on anything else locally
+#
+
+# build man pages for externally available interfaces
+
+# for each header file, get a list of man pages
+# from that file, then process each page, one
+# at a time
+foreach api_file: libiscsi_usr_doc_headers
+  # get a list of pages for this header file
+  c = run_command(list_pages_cmd, api_file, check: true)
+  man_pages = c.stdout().split()
+  # process each page, one at a time, generating dependencies
+  foreach page: man_pages
+    custom_target(
+      page.underscorify() + '_man',
+      input: api_file,
+      output: page + '.3',
+      capture: true,
+      command: [
+        kernel_doc_cmd,
```

Comment:
https://github.com/open-iscsi/open-iscsi/blob/7726ecc89015edb49057473a605c63ac9354f5cb/libopeniscsiusr/docs/kernel-doc#L45-L48

Your kernel_doc_cmd prints the file to stdout, and in order for Meson to save that to a file it needs to run `meson --internal exe --capture page.3 kernel-doc -module libopeniscsiusr -man -function ...`

So, it's running a python program that runs a perl program, and the only thing the python program does is read stdout and save it to a file.

IMHO it would be more efficient to just add an `-outfile` argument to the kernel-doc script so that instead of doing `print "...";` everywhere, you can just print directly to the intended file.

Reply:
I really don't want to mess with kernel-doc. It's an established (if largely hacked) perl script from the kernel (obviously). Could I modify perl and add an output argument? Probably, with a bit of time and web searching. But if the only cost of not doing so is less efficient man page generation, I'll live with it.

Frankly, it seems like a lot of work for nothing, since the interfaces rarely change or get added to. I'd just use static man pages if it was me. :) But I didn't write that code, so I'm trying to just keep it working.

I might look at improving that code, but I'll do so as a separate issue.

Reply:
Alright, cool. Just wanted to give you a heads up. I mainly.mention it because given a choice, it's "cleaner" to do it with integrated handling, and many people who use it simply don't realize it matters at all.

It's a bit slower but not terrible except on Windows where forking additional programs is an actually noticeable problem, and that's, uh. Not a problem here. So, you're well warranted to say it's not worth the effort to rewrite someone else's borrowed script just for a bit of cleanup and micro-optimizing the already significantly faster build time.